### PR TITLE
SkSL precompile

### PR DESCRIPTION
**Pull request by liyuqian**
For https://github.com/flutter/flutter/issues/40686

Unit tests added:
- CacheSkSLWorks
- VisitFilesCanBeCalledTwice
- CanListFilesRecursively

### CODE REVIEW

**Comment by liyuqian** [comment]
CC @iskakaushik : just in case that you're interested in the shader warm-up work. Would you like to review this when it's out of wip?

**Comment by iskakaushik** [comment]
@liyuqian yes please!

File: fml/platform/posix/file_posix.cc [hunk 1]
```diff
@@ -210,4 +211,20 @@ bool WriteAtomically(const fml::UniqueFD& base_directory,
...
+  std::vector<std::string> files;
+  DIR* dir = ::fdopendir(directory.get());
+  if (dir == nullptr) {
+    FML_LOG(ERROR) << "Can't open the directory.";
+    return files;
+  }
+  while (dirent* ent = readdir(dir)) {
+    std::string name = ent->d_name;
+    if (name != "." && name != "..") {
+      files.push_back(std::move(name));
```
Comment (chinmaygarde) [inline fml/platform/posix/file_posix.cc]: Optional suggestion: Consider using a visitor pattern here. For example, `size_t ListFiles(const fml::UniqueFD& directory, bool(const std::string& file) file_callback)` should work and is way more flexible IMO. With this in place, you can also make this call recursive if the directory entry is itself a directory. The `bool` return from the visitor callback can be an indication to the caller about whether further traversal is necessary.

File: fml/platform/posix/file_posix.cc [hunk 2]
```diff
@@ -210,4 +211,20 @@ bool WriteAtomically(const fml::UniqueFD& base_directory,
                     base_directory.get(), file_name) == 0;
 }
 
+std::vector<std::string> ListFiles(const fml::UniqueFD& directory) {
+  std::vector<std::string> files;
+  DIR* dir = ::fdopendir(directory.get());
+  if (dir == nullptr) {
+    FML_LOG(ERROR) << "Can't open the directory.";
```
Comment (chinmaygarde) [inline fml/platform/posix/file_posix.cc]: Nit: Also log strerror(errno) so it is easier to tell why the directory could not be opened. Also, maybe this should be a debug log?

**Comment by liyuqian** [comment]
@chinmaygarde @iskakaushik : unit tests are added and the PR is now ready for review.

File: fml/file.h [hunk 3]
```diff
@@ -73,12 +78,37 @@ bool WriteAtomically(const fml::UniqueFD& base_directory,
                      const char* file_name,
                      const Mapping& mapping);
 
+using FileVisitor = std::function<void(const fml::UniqueFD& directory,
```
Comment (chinmaygarde) [inline fml/file.h]: Can we make the `FileVisitor` return a `bool` to indicate if further traversal is necessary? For example, folks can use this to implement a `FindFile` whereby they can return `false` when a file is found so further traversal is unnecessary.

File: fml/file.h [hunk 4]
```diff
@@ -73,12 +78,37 @@ bool WriteAtomically(const fml::UniqueFD& base_directory,
...
+                                       const std::string& filename)>;
+
+/// Call `visitor` on all files inside the `directory` non-recursively. The
+/// trivial file "." and ".." will not be visited.
+///
+/// If recursive visiting is needed, call `VisitFiles` inside the `visitor`, or
+/// use our helper method `VisitFilesRecursively`.
+///
+/// @see `VisitFilesRecursively`.
+void VisitFiles(const fml::UniqueFD& directory, const FileVisitor& visitor);
```
Comment (chinmaygarde) [inline fml/file.h]: You don't need to pass closures by const reference. Just `FileVisitor visitor` is fine.

File: fml/file.h (hunk 3 above)
Comment (chinmaygarde) [reply fml/file.h]: Please document if the filename is relative or absolute.

File: fml/platform/posix/file_posix.cc [hunk 5]
```diff
@@ -210,4 +211,20 @@ bool WriteAtomically(const fml::UniqueFD& base_directory,
                     base_directory.get(), file_name) == 0;
 }
 
+std::vector<std::string> ListFiles(const fml::UniqueFD& directory) {
+  std::vector<std::string> files;
+  DIR* dir = ::fdopendir(directory.get());
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply fml/platform/posix/file_posix.cc]: `::closedir` won't close the FD. The `FD` and the `DIR*` are different.  Maybe that is why you had to `rewinddir`?

**Comment by liyuqian** [comment]
Calling `::closedir` will make `FileTest.VisitFilesCanBeCalledTwice` fail with
```
[ERROR:flutter/fml/platform/posix/file_posix.cc(226)] Can't open the directory. Error: Bad file descriptor
../../flutter/fml/file_unittest.cc:154: Failure
```

BTW, I've also reset your accidental approval :)

File: fml/file.cc [hunk 6]
```diff
@@ -58,4 +59,23 @@ ScopedTemporaryDirectory::~ScopedTemporaryDirectory() {
   }
 }
 
+void VisitFilesRecursively(const fml::UniqueFD& directory,
+                           const FileVisitor& visitor) {
+  FileVisitor recursive_visitor = [&recursive_visitor, &visitor](
```
Comment (iskakaushik) [inline fml/file.cc]: nit: you could use nftw https://linux.die.net/man/3/nftw to simplify this.

File: fml/file.h (hunk 3 above)
Comment (iskakaushik) [reply fml/file.h]: `filename` inherently implies it's the name of the file as opposed to filepath, no? :-)

File: shell/common/persistent_cache.h [hunk 7]
```diff
@@ -49,25 +49,48 @@ class PersistentCache : public GrContextOptions::PersistentCache {
...
+
+  using SkSLCache = std::pair<sk_sp<SkData>, sk_sp<SkData>>;
+
+  /// Load all the SkSL shader caches in the right directory.
+  std::vector<SkSLCache> LoadSkSLs();
+
+  static bool cache_sksl() { return cache_sksl_; }
+  static void SetCacheSkSL(bool value);
+  static void MarkStrategySet() { strategy_set_ = true; }
+
```
Comment (iskakaushik) [inline shell/common/persistent_cache.h]: I wonder if we can use some builder like pattern here. Like:

```
PersistentCacheBuilder
  .setCacheSkSl(true)
  ...
  .build()
```

That way we could avoid having this extra flag. But i'm not sure if this is a common c++ paradigm.

Comment (chinmaygarde) [reply shell/common/persistent_cache.h]: We use this pattern in the `embedder_unittests` in for the `EmbedderConfigBuilder`.

File: fml/file.h (hunk 4 above)
Comment (liyuqian) [reply fml/file.h]: Done.

File: fml/platform/posix/file_posix.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (liyuqian) [reply fml/platform/posix/file_posix.cc]: See https://github.com/flutter/engine/pull/12412#issuecomment-537261733

File: shell/common/persistent_cache.h (hunk 7 above)
Comment (liyuqian) [reply shell/common/persistent_cache.h]: I like the builder pattern, but it doesn't seem to be compatible with our current ` PersistentCache::GetCacheForProcess` usages. If we think that it's worth to do a yak shave to remove ` PersistentCache::GetCacheForProcess` and let `Shell` own the `PersistentCache` used by `Rasterizer` and `IOManager`, maybe we shall do that in another isolated PR and create an issue to track its progress.

**Comment by liyuqian** [comment]
I've finally made all tests pass on both Windows and Linux 😄  This is now ready for review again @chinmaygarde @iskakaushik

**Pull request merged**